orchestrators/Orchestrator.py

```diff
@@ -0,0 +1,443 @@
+    ELASTICSEARCH_FILENAME_COLUMN = os.environ.get("ELASTICSEARCH_FILENAME_COLUMN")
+    ELASTICSEARCH_TITLE_COLUMN = os.environ.get("ELASTICSEARCH_TITLE_COLUMN")
+    ELASTICSEARCH_URL_COLUMN = os.environ.get("ELASTICSEARCH_URL_COLUMN")
+    ELASTICSEARCH_VECTOR_COLUMNS = os.environ.get("ELASTICSEARCH_VECTOR_COLUMNS")
+    ELASTICSEARCH_STRICTNESS = os.environ.get("ELASTICSEARCH_STRICTNESS", SEARCH_STRICTNESS)
+    ELASTICSEARCH_EMBEDDING_MODEL_ID = os.environ.get("ELASTICSEARCH_EMBEDDING_MODEL_ID")
+
+    SHOULD_STREAM = True if AZURE_OPENAI_STREAM.lower() == "true" else False
+
+    # methods to implement in orchestrator
+    def fetchUserGroups(self, userToken, nextLink=None):
+        # Recursively fetch group membership
+        if nextLink:
+            endpoint = nextLink
+        else:
+            endpoint = "https://graph.microsoft.com/v1.0/me/transitiveMemberOf?$select=id"
+        
+        headers = {
+            'Authorization': "bearer " + userToken
+        }
+        try :
+            r = requests.get(endpoint, headers=headers)
+            if r.status_code != 200:
+                if self.DEBUG_LOGGING:
+                    logging.error(f"Error fetching user groups: {r.status_code} {r.text}")
+                return []
+            
+            r = r.json()
+            if "@odata.nextLink" in r:
+                nextLinkData = self.fetchUserGroups(userToken, r["@odata.nextLink"])
+                r['value'].extend(nextLinkData)
+            
+            return r['value']
+        except Exception as e:
+            logging.error(f"Exception in fetchUserGroups: {e}")
+            return []
+
+    # Filter for permitted user groups    
+    def generateFilterString(self, userToken):
+        # Get list of groups user is a member of
+        userGroups = self.fetchUserGroups(self, userToken)
+
+        # Construct filter string
+        if not userGroups:
+            logging.debug("No user groups found")
+
+        group_ids = ", ".join([obj['id'] for obj in userGroups])
+        return f"{self.AZURE_SEARCH_PERMITTED_GROUPS_COLUMN}/any(g:search.in(g, '{group_ids}'))"
+
+    # Format response as newline delimited json
+    def format_as_ndjson(self, obj: dict) -> str:
+        return json.dumps(obj, ensure_ascii=False) + "\n"
+
+    # Format request body and headers with relevant info based on search type
+    def prepare_body_headers_with_data(self, request):
+        request_messages = request.json["messages"]
+
+        body = {
+            "messages": request_messages,
+            "temperature": float(self.AZURE_OPENAI_TEMPERATURE),
+            "max_tokens": int(self.AZURE_OPENAI_MAX_TOKENS),
+            "top_p": float(self.AZURE_OPENAI_TOP_P),
+            "stop": self.AZURE_OPENAI_STOP_SEQUENCE.split("|") if self.AZURE_OPENAI_STOP_SEQUENCE else None,
+            "stream": self.SHOULD_STREAM,
+            "dataSources": []
+        }
+
+        if self.DATASOURCE_TYPE == "AzureCognitiveSearch":
```

Comment:
Consider if an enum might be appropriate for this check instead of a magic string.